streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -375,42 +341,61 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
         updateBufferMetrics();
     }
 
+    private byte[] inferPriorValue(final Bytes key, final byte[] serializedValue) {
+        final byte[] priorValue;
+        if (index.containsKey(key)) {
+            priorValue = internalPriorValueForBuffered(key);
+        } else {
+            final Change<V> change = valueSerde.deserializer().deserialize(
```

Comment:
I don't follow what you're doing here. In the case we're just loading it into the buffer from the changelog for the first time, we actually do just want oldValue to be null right? We shouldn't need to deserialize at all

Reply:
no, when we're observing a key from the changelog that's not already buffered, it's the same as if the processor is buffering a key for the first time. The `oldValue` is set to whatever the previous state of the KTable was before the change we're suppressing. As of the instant before we suppress an update, the parent KTable and the suppressed KTable have the same view of the record. This is why (in both restoration and in processing) we can interpret the oldValue of the first change as the "prior state".

Does that make sense? It sort of still feels confusing, but I'm for some reason finding it hard to express this idea clearly :/

Reply:
After discussing this I'm in agreement here. It's helpful to note that the serializedValue is in fact a serialized Change -- not sure if renaming this would lessen or add to the confusion...

Reply:
Hmm, tentatively, I think that renaming wouldn't be possible. The field in question is the "value" field of ContextualRecord, which is only always a change in here because that's what we put in there. Elsewhere in Streams, it can be anything in the "value" field.

Reply:
I meant just the local variables here, not the field itself

Reply:
Ah. Sorry. I've made the change.